source/postprocess/heating_statistics.cc

```diff
@@ -0,0 +1,177 @@
+
+      double average_heating_integral = 0.0;
+      double total_heating_integral = 0.0;
+
+      unsigned int index = 0;
+      for (typename std::list<std_cxx11::shared_ptr<HeatingModel::Interface<dim> > >::const_iterator
+           heating_model = heating_model_objects.begin();
+           heating_model != heating_model_objects.end(); ++heating_model, ++index)
+        {
+          typename DoFHandler<dim>::active_cell_iterator
+          cell = this->get_dof_handler().begin_active(),
+          endc = this->get_dof_handler().end();
+
+          double local_heating_integrals = 0;
+          double local_mass = 0;
+
+          // compute the integral quantities by quadrature
+          for (; cell!=endc; ++cell)
+            if (cell->is_locally_owned())
+              {
+                fe_values.reinit (cell);
+                fe_values[this->introspection().extractors.temperature]
+                .get_function_values (this->get_solution(),
+                                      in.temperature);
+                fe_values[this->introspection().extractors.pressure]
+                .get_function_values (this->get_solution(),
+                                      in.pressure);
+                fe_values[this->introspection().extractors.velocities]
+                .get_function_values (this->get_solution(),
+                                      in.velocity);
+                fe_values[this->introspection().extractors.pressure]
+                .get_function_gradients (this->get_solution(),
+                                         in.pressure_gradient);
+                for (unsigned int c=0; c<this->n_compositional_fields(); ++c)
+                  fe_values[this->introspection().extractors.compositional_fields[c]]
+                  .get_function_values(this->get_solution(),
+                                       composition_values[c]);
+                for (unsigned int i=0; i<fe_values.n_quadrature_points; ++i)
+                  {
+                    for (unsigned int c=0; c<this->n_compositional_fields(); ++c)
+                      in.composition[i][c] = composition_values[c][i];
+                  }
+
+                in.position = fe_values.get_quadrature_points();
+
+                this->get_material_model().evaluate(in, out);
+                (*heating_model)->evaluate(in, out, heating_model_outputs);
+
+                for (unsigned int q=0; q<n_q_points; ++q)
+                  {
+                    local_heating_integrals += heating_model_outputs.heating_source_terms[q]
+                                                        * out.densities[q] * fe_values.JxW(q);
+
+                    local_mass += out.densities[q] * fe_values.JxW(q);
+                  }
+              }
+
+          // compute the sum over all processors
+          std::vector<double> local_value;
+          std::vector<double> global_value;
+          local_value.push_back(local_heating_integrals);
+          local_value.push_back(local_mass);
```

Comment:
Since you use an array of exactly two values, can't you just use
```
  const double local_values[2] = { local_heating_integrals, local_mass };
  double global_values[2];
```
?